File: Stack/MinStack.py
class MinStack(object):

    def __init__(self):
        self.s = []
        self.mins = []

    def push(self, val):
        self.s.append(val)
        if len(self.s) == 1:
            self.mins.append(val)
        else:
            if self.mins[len(self.mins)-1] < val:
                self.mins.append(self.mins[len(self.mins)-1])
            else:
                self.mins.append(val)

    def pop(self):
        self.s.pop()
        self.mins.pop()
        

    def top(self):
        return self.s[len(self.s)-1]
        

    def getMin(self):
        return self.mins[len(self.mins)-1]

File: Stack/MinStack.cpp
#include <vector>
#include <string>

using namespace std;

class MinStack {
public:
    MinStack() {}
    
    void push(int val) {
        s.push_back(val);
        if(s.size() == 1)
            mins.push_back(val);
        else
        {
            if(mins[mins.size()-1] < val)
                mins.push_back(mins[mins.size()-1]);
            else
                mins.push_back(val);
        }
    }
    
    void pop() {
        s.pop_back();
        mins.pop_back();
    }
    
    int top() {
        return s[s.size()-1];
    }
    
    int getMin() {
        return mins[s.size()-1];
    }
private:
    vector<int> s;
    vector<int> mins;
};

